### PR TITLE
Implement optimized median filter

### PR DESCRIPTION
**Pull request by camillescott**
Easy optimization to improve normalize-by-median: it's simple to observe that checking if the median of a set is greater than some cutoff is equivalent to checking if more than half the elements of that set are greater than some cutoff. The latter avoids doing a lookup for every kmer every time, and avoids a costly sort. On a small dataset (1m ecoli reads), this was an 18% performance improvement. Not bad for 5 minutes!

TODO: Add a couple tests

### CODE REVIEW

**Comment by mr-c** [comment]
Cool!

**Comment by camillescott** [comment]
- [ ] Is it mergeable?
   - [ ] Did it pass the tests?
   - [ ] If it introduces new functionality in scripts/ is it tested?
     Check for code coverage with `make clean diff-cover`
   - [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
     `make cppcheck`, and `make doc` output. Use `make format` and manual
     fixing as needed.
   - [ ] Did it change the command-line interface? Only additions are allowed
     without a major version increment. Changing file formats also requires a
     major version number increment.
   - [ ] Is it documented in the ChangeLog?
     http://en.wikipedia.org/wiki/Changelog#Format
   - [ ] Was a spellchecker run on the source code and documentation after
     changes were made?

**Comment by camillescott** [comment]
ready for review @ctb @mr-c

**Comment by ctb** [comment]
Does this return identical results to old approach?

**Comment by camillescott** [comment]
Yup. All normalize-by-median tests pass with the new function patched in.

**Comment by camillescott** [comment]
jenkins, test this please you useless drunk

**Comment by camillescott** [comment]
jenkins test this please and give it a pass or the floggings will continue

**Comment by camillescott** [comment]
@ctb @mr-c @luizirber okey now pls review

**Comment by ctb** [comment]
Please leave for me to merge - tnx.

**Comment by camillescott** [comment]
@ctb I keep expecting that I've overlooked something really obvious and there's no way a 20% performance increase would be so easy, but *shrug*

File: khmer/_khmermodule.cc [hunk 1]
```diff
@@ -938,6 +938,33 @@ hash_get_max_count(khmer_KCountingHash_Object * me, PyObject * args)
...
+        return NULL;
+    }
+
+    if (strlen(long_str) < counting->ksize()) {
+        PyErr_SetString(PyExc_ValueError,
+                        "string length must >= the hashtable k-mer size");
+        return NULL;
+    }
+
+    if (counting->filter_on_median(long_str, cutoff))
```
Comment (ctb) [inline khmer/_khmermodule.cc]: Please add { and } - single-line if statements are dangerous ;).

File: lib/hashtable.hh [hunk 2]
```diff
@@ -291,6 +291,9 @@ public:
         unsigned long long  &n_consumed
     );
 
+    bool filter_on_median(const std::string &s,
+                        unsigned int cutoff);
```
Comment (ctb) [inline lib/hashtable.hh]: spacing should be adjusted...

File: lib/hashtable.cc [hunk 3]
```diff
@@ -231,6 +231,29 @@ void Hashtable::get_median_count(const std::string &s,
...
+//
+// Optimized filter function for normalize-by-median:
+// why do all the count lookups and an expensive sort
+// every time when we can just filter if more than half the
+// kmer counts are above the cutoff?
+//
+bool Hashtable::filter_on_median(const std::string &s,
+                                unsigned int cutoff) {
+    KMerIterator kmers(s.c_str(), _ksize);
+    unsigned int min_req = (s.size() - _ksize + 1) / 2;
```
Comment (ctb) [inline lib/hashtable.cc]: Please round up - (0.5 + float(s.size() - _ksize + 1) / 2.0);

**Comment by ctb** [comment]
A few comments --

the code didn't return identical results on data/100k-filtered.fa, due to a round-off error.  See rounding comment on hashtable.cc.

Please add tests for single k-mer, med < cutoff; single k-mer, med > cutoff; odd # k-mers, med < cutoff; odd # of k-mers; med > cutoff; and any other edge cases you can think of (that's all I got).  The homopolymer run tests are not sufficient ;).  You should enforce comparison with the old function, too, to make sure they both return the same results.

can we implement the old function without the sort? it seems fairly straightforward to do so although there would still be a performance hit.  This shouldn't be your job to do, but can you create a few issues around (a) improving old function, and (b) replacing old function with this function in places where it would work?

tnx!

**Comment by camillescott** [comment]
@ctb can do on the extra tests and whatnot.

We definitely can't do the old function without the sort -- the key difference here is that the actual median isn't calculated, only that it's greater than some value. There are definitely more efficient median algorithms if we want to replace the naive sort-based one we're using now thoug

**Comment by ctb** [comment]
On Mon, Mar 09, 2015 at 01:15:38PM -0700, Camille Scott wrote:
> @ctb can do on the extra tests and whatnot.

ok, let me know

> We definitely can't do the old function without the sort -- the key difference here is that the actual median isn't calculated, only that it's greater than some value. There are definitely more efficient median algorithms if we want to replace the naive sort-based one we're using now thoug

sure, we should look at the current code and think about where we might
optimize.

**Comment by camillescott** [comment]
Stop being useless Jenkins, and dammit Jenkins, test this please!

**Comment by camillescott** [comment]
Stop being useless Jenkins, and dammit Jenkins, test this please!

**Comment by camillescott** [comment]
Stop being useless Jenkins, and dammit Jenkins, test this please!

**Comment by camillescott** [comment]
@ctb, ready for review

File: khmer/_khmermodule.cc (hunk 1 above)
Comment (ctb) [inline khmer/_khmermodule.cc]: Ping.

File: tests/test_counting_hash.py [hunk 4]
```diff
@@ -212,6 +212,135 @@ def test_simple_median():
...
+            'GTACGTACGTACGTACGTACCCC',
+            'TTAGTTAGTTAGTTAGTTAGCCC']
+
+    for seq in seqs:
+        hi.consume(seq)
+        hi.consume(seq)
+        hi.consume(seq)
+
+        med, _, _ = hi.get_median_count(seq)
+        assert hi.filter_on_median(seq, 4) is (med >= C)
```
Comment (ctb) [inline tests/test_counting_hash.py]: replace 4 with C!

File: tests/test_counting_hash.py [hunk 5]
```diff
@@ -212,6 +212,135 @@ def test_simple_median():
...
+    seqs = ['ATCGATCGATCGATCGATCGCCC',
+            'GTACGTACGTACGTACGTACCCC',
+            'TTAGTTAGTTAGTTAGTTAGCCC']
+
+    for seq in seqs:
+        hi.consume(seq)
+        assert hi.filter_on_median(seq, 1) is True
+
+
+def test_filter_on_median_even_lt():
```
Comment (ctb) [inline tests/test_counting_hash.py]: please annotate the difference from 'test_filter_on_median_odd_lt' (seqs are different lengths)

**Comment by ctb** [comment]
Two questions for consideration cc @mr-c --

* should we store the results of normalize-by-median on (say) 100k-filtered.fa, as a test for consistency going forward?

* filter_on_median is an unsatisfying function name, since it doesn't actually filter anything. Would 'median_greater_than' or something like that be clearer? Suggestions welcome.

**Comment by mr-c** [comment]
+1 for storing known-good output. How big, when compressed? Worse case we keep it elsewhere and let Jenkins deal with it.

**Comment by camillescott** [comment]
Very well -- I'll add a comparison test to known output of the mentioned file and change the function name.

**Comment by camillescott** [comment]
@ctb @mr-c okay I added the new test -- it adds ~5mb the repo. Is that manageable?

**Comment by mr-c** [comment]
Hrmm.. that would nearly double the distribution size..

**Comment by camillescott** [comment]
@ctb @mr-c further thoughts on this??

**Comment by mr-c** [comment]
We need a separate repo to store larger pieces of test data and have tests that run only on Jenkins by default. @ctb What do you think?

**Comment by bocajnotnef** [comment]
Would something like git's LFS work?

https://github.com/blog/1986-announcing-git-large-file-storage-lfs

**Comment by mr-c** [comment]
@bocajnotnef In this case the file is 5Megabytes so GitHub LFS wouldn't be needed. I don't want to make the git clone of the main repository larger than necessary. Another option is to add the 5M file but exclude it from packaging and adjust the test to skip if the file isn't present.

**Comment by ctb** [comment]
On Thu, Apr 09, 2015 at 08:55:01AM -0700, Michael R. Crusoe wrote:
> @bocajnotnef In this case the file is 5Megabytes so GitHub LFS wouldn't be needed. I don't want to make the git clone of the main repository larger than necessary. Another option is to add the 5M file but exclude it from packaging and adjust the test to skip if the file isn't present.

What about a 'khmer-data' repository?  Although that seems kind of silly.

**Comment by ctb** [comment]
On Thu, Apr 09, 2015 at 08:19:19AM -0700, Michael R. Crusoe wrote:
> We need a separate repo to store larger pieces of test data and have tests that run only on Jenkins by default. @ctb What do you think?

Since the data is mostly going to be read-only, it seems silly to have a full
repo.  Maybe their new data extension *is* the way to go for that?  And have a
khmer-data repo that stores the pointers?

**Comment by camillescott** [comment]
How about we get the optimization merged for now, then open an issue for supporting github LFS or some other variety of data hosting to be addressed in the future?

**Comment by mr-c** [comment]
+1

**Comment by ctb** [comment]
- [ ] create an issue
- [ ] make mergable

**Comment by ctb** [comment]
@camillescott - @mr-c is going to address data size problem for test by putting that elsewhere, but other than that, all you need to do is make it mergable, I believe.

**Comment by mr-c** [comment]
It can go into https://github.com/dib-lab/khmer-testdata ; just commit directly

**Comment by ctb** [comment]
ping @camillescott

**Comment by camillescott** [comment]
@mr-c need access to khmer-testdata repo

**Comment by camillescott** [comment]
jenkins, test this please

**Comment by Blahah** [comment]
There is so much winning in this PR, @camillescott.

**Comment by ctb** [comment]
Suggest squashing as in #660.

**Comment by camillescott** [comment]
Jenkins, test this please

**Comment by camillescott** [comment]
@ctb ready for final review / merge

**Comment by camillescott** [comment]
also pleeeeeeeeeease merge this before merging any other PR's in, it's all squashed and pretty and I've had to update and tweak little things on this branch way too many times :'(

**Comment by ctb** [comment]
LGTM; nice work.

**Comment by camillescott** [comment]
Yay, thanks!

File: lib/hashtable.cc [hunk 6]
```diff
@@ -232,6 +232,26 @@ void Hashtable::get_median_count(const std::string &s,
...
+bool Hashtable::median_at_least(const std::string &s,
+                                unsigned int cutoff) {
+    KMerIterator kmers(s.c_str(), _ksize);
+    unsigned int min_req = 0.5 + float(s.size() - _ksize + 1) / 2;
+    unsigned int num_cutoff_kmers = 0;
+    while(!kmers.done()) {
+        HashIntoType kmer = kmers.next();
+        if (this->get_count(kmer) >= cutoff) {
+            ++num_cutoff_kmers;
+            if (num_cutoff_kmers >= min_req) {
```
Comment (ctb) [inline lib/hashtable.cc]: Random thought @camillescott - what if you did a for or a while loop up until min_req, and only then checked the 'if'? Obviously if num_kmers < min_req there's no point in checking.

Comment (camillescott) [inline lib/hashtable.cc]: :cries: that's probably fair, though I imagine the compiler + cpu does a fair job with it's branch prediction in this scenario

Comment (ctb) [inline lib/hashtable.cc]: On Tue, Jun 09, 2015 at 10:20:17AM -0700, Camille Scott wrote:
> > @@ -232,6 +232,26 @@ void Hashtable::get_median_count(const std::string &s,
> >      median = counts[counts.size() / 2]; // rounds down
> >  }
> >  
> > +//
> > +// Optimized filter function for normalize-by-median
> > +//
> > +bool Hashtable::median_at_least(const std::string &s,
> > +                                unsigned int cutoff) {
> > +    KMerIterator kmers(s.c_str(), _ksize);
> > +    unsigned int min_req = 0.5 + float(s.size() - _ksize + 1) / 2;
> > +    unsigned int num_cutoff_kmers = 0;
> > +    while(!kmers.done()) {
> > +        HashIntoType kmer = kmers.next();
> > +        if (this->get_count(kmer) >= cutoff) {
> > +            ++num_cutoff_kmers;
> > +            if (num_cutoff_kmers >= min_req) {
> 
> :cries: that's probably fair, though I imagine the compiler + cpu does a fair job with it's branch prediction in this scenario

:)
